src/base/iterators.jl

```diff
@@ -1,39 +1,52 @@
-function Base.iterate(layer::AbstractFeatureLayer, state::Int=0)
+function Base.iterate(
+    layer::AbstractFeatureLayer,
+    state::Integer = 0,
+)::Union{Nothing,Tuple{Feature,Int64}}
     layer.ptr == C_NULL && return nothing
     state == 0 && resetreading!(layer)
     ptr = GDAL.ogr_l_getnextfeature(layer.ptr)
-    if ptr == C_NULL
+    return if ptr == C_NULL
         resetreading!(layer)
-        return nothing
+        nothing
     else
-        return (Feature(ptr), state+1)
+        (Feature(ptr), state + 1)
     end
 end
 
-Base.eltype(layer::AbstractFeatureLayer) = Feature
+Base.eltype(layer::AbstractFeatureLayer)::DataType = Feature
 
-Base.length(layer::AbstractFeatureLayer) = nfeature(layer, true)
+Base.length(layer::AbstractFeatureLayer)::Integer = nfeature(layer, true)
 
-struct BlockIterator
-    rows::Cint
-    cols::Cint
-    ni::Cint
-    nj::Cint
-    n::Cint
-    xbsize::Cint
-    ybsize::Cint
+struct BlockIterator{T<:Integer}
+    rows::T
+    cols::T
+    ni::T
+    nj::T
+    n::T
+    xbsize::T
+    ybsize::T
 end
 
-function blocks(raster::AbstractRasterBand)
+function blocks(
+    ::Type{T},
+    raster::AbstractRasterBand,
+)::BlockIterator{T} where {T<:Integer}
     (xbsize, ybsize) = blocksize(raster)
-    rows = height(raster)
-    cols = width(raster)
-    ni = ceil(Cint, rows / ybsize)
-    nj = ceil(Cint, cols / xbsize)
-    BlockIterator(rows, cols, ni, nj, ni * nj, xbsize, ybsize)
+    rows = T(height(raster))
+    cols = T(width(raster))
+    ni = ceil(T, rows / ybsize)
+    nj = ceil(T, cols / xbsize)
+    return BlockIterator{T}(rows, cols, ni, nj, ni * nj, xbsize, ybsize)
 end
 
-function Base.iterate(obj::BlockIterator, iter::Int=0)
+function blocks(raster::AbstractRasterBand)::BlockIterator{Int64}
+    return blocks(Int64, raster)
+end
+
+function Base.iterate(
+    obj::BlockIterator{T},
+    iter::T = 0,
+)::Union{Nothing,Tuple{Tuple{Tuple{T,T},Tuple{T,T}},T}} where {T}
```

Comment:
That's one of the more interesting types I've seen written out.

Reply:
😛